File: python/1011. Capacity To Ship Packages Within D Days/solution.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        def carriableInDays(weights, capacity, days):
            daysNeeded = 0
            currentLoad = 0
            for i in range(len(weights)):
                if (currentLoad + weights[i] > capacity):
                    currentLoad = weights[i]
                    daysNeeded += 1
                else:
                    currentLoad += weights[i]

            if currentLoad > 0:
                daysNeeded += 1

            return daysNeeded <= days

        left = 0
        right = 0

        for weight in weights:
            left = max(left, weight)
            right += weight

        while left < right:
            mid = left + (right - left) // 2

            if carriableInDays(weights, mid, days):
                right = mid
            else:
                left = mid + 1

        return left

File: java/1011. Capacity To Ship Packages Within D Days/Solution.java
class Solution {
    // min capacity = max weight
    // max capactiy = total weight
    // search between these two values
    public int shipWithinDays(int[] weights, int days) {
        int totalWeight = 0;
        int biggest = 0;
        for (int weight : weights) {
            totalWeight += weight;
            biggest = Math.max(biggest, weight);
        }

        int left = biggest;
        int right = totalWeight;

        while (left < right) {
            int mid = left + (right - left) / 2;
            int daysNeeded = calculateDaysNeeded(mid, weights);
            if (daysNeeded > days) left = mid + 1;
            if (daysNeeded <= days) right = mid;
        }
        return left;
    }

    public int calculateDaysNeeded(int capacity, int[] weights) {
        int days = 0;
        int currentLoad = 0;
        for (int i = 0; i < weights.length; i++) {
            if (currentLoad + weights[i] <= capacity) {
                currentLoad += weights[i];
            } else {
                days++;
                currentLoad = weights[i];
            }
        }
        return currentLoad > 0 ? ++days : days;
    }
}